x-pack/legacy/plugins/ml/public/application/jobs/components/custom_url_editor/utils.js

```diff
@@ -152,52 +156,42 @@ function buildDashboardUrlFromSettings(settings) {
           query = searchSourceData.query;
         }
 
-        // Add time settings to the global state URL parameter with $earliest$ and
-        // $latest$ tokens which get substituted for times around the time of the
-        // anomaly on which the URL will be run against.
-        const _g = rison.encode({
-          time: {
-            from: '$earliest$',
-            to: '$latest$',
-            mode: 'absolute',
-          },
-        });
-
-        const appState = {
-          filters,
-        };
-
-        // To put entities in filters section would involve creating parameters of the form
-        // filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:b30fd340-efb4-11e7-a600-0f58b1422b87,
-        // key:airline,negate:!f,params:(query:AAL,type:phrase),type:phrase,value:AAL),query:(match:(airline:(query:AAL,type:phrase)))))
-        // which includes the ID of the index holding the field used in the filter.
-
-        // So for simplicity, put entities in the query, replacing any query which is there already.
-        // e.g. query:(language:kuery,query:'region:us-east-1%20and%20instance:i-20d061fa')
         const queryFromEntityFieldNames = buildAppStateQueryParam(queryFieldNames);
         if (queryFromEntityFieldNames !== undefined) {
           query = queryFromEntityFieldNames;
         }
 
-        if (query !== undefined) {
-          appState.query = query;
-        }
-
-        const _a = rison.encode(appState);
-
-        const urlValue = `kibana#/dashboard/${dashboardId}?_g=${_g}&_a=${_a}`;
-
-        const urlToAdd = {
-          url_name: settings.label,
-          url_value: urlValue,
-          time_range: TIME_RANGE_TYPE.AUTO,
-        };
-
-        if (settings.timeRange.type === TIME_RANGE_TYPE.INTERVAL) {
-          urlToAdd.time_range = settings.timeRange.interval;
-        }
+        const generator = npStart.plugins.directAccessLinks.getAccessLinkGenerator(
+          DASHBOARD_APP_LINK_GENERATOR
+        );
+
+        return generator
+          .createUrl({
+            dashboardId,
+            timeRange: {
+              from: '$earliest$',
+              to: '$latest$',
+              mode: 'absolute',
+            },
+            filters,
+            query,
+            // Don't hash the URL since this string will be 1. shown to the user and 2. used as a
+            // template to inject the time parameters.
+            useHash: false,
+          })
+          .then(urlValue => {
```

Comment:
@peteharverson - would a service like this benefit ML?  Unfortunately I had to strip the base path and the `app` and I manually re-added `kibana` prefix, since that's what you show users in the input box.

This would buffer the ml team at least from changes to the hash that dashboard app team may make.  It looks like though you store the relative urls in saved objects?  Have you thought about how to deal with migrations? in https://github.com/elastic/kibana/issues/25247 I discuss some issues with storing urls as string in saved objects.  Ideally if we had this service you could store the generator id and the state and generate the url string on the fly, but since you expose this url string to the user, that may not be an option for you.

Reply:
@stacey-gammon yes ML would definitely benefit from having some sort of service to generate links to Discover and Dashboard, and indeed to any other apps such as APM, SIEM, Logs UI. The current process by which we build the URLs to Discover / Dashboard is cumbersome, and if the format of those URLs change (which to be fair hasn't happened over previous releases) then we need to edit the URL building code inside ML and all the custom URLs of ML jobs inside our pre-built modules.

When generating drilldowns from ML jobs, the properties we would need to store in the state are:
 - dashboard ID / discover index pattern ID
 - time range to set in the target page
 - entities to pass to Discover / Dashboard to set in the 'filter', for example 'user: jane' and 'host: server1'. Currently we add these to the query part of the URL, as it is much easier to add them to a query than to the filter part. Not sure if we would want to allow the user to specify whether these entities should be set in the query or the filter - it might make more sense for the target app to decide how to use these filter entities?

The main reason we display the actual URL to the user in the edit job flyout / jobs list is historical. If we switched to this new service, we could just display controls for the state (saved object ID, time range, entities etc) rather than the full URL. This would also make it easier for the user to edit the custom URL, rather than the current method where they have to directly edit the long URL string, which is error-prone. We would then generate the actual URL on the fly, when the user clicked on the action link in the anomalies table for example. If a user for whatever reason couldn't use one of these new format URLs, then they could always use the `other` type of custom URL which allows them to enter the URL directly.

Note the custom URL links are stored within ML jobs (documents in ES, stored in the `.ml-config` index and accessed via `GET _ml/anomaly_detectors`), and not in Kibana saved objects. As part of the upcoming work to make ML jobs space-aware, we are planning on adding saved object 'wrappers' around ML jobs, so in theory, we could move the custom URLs into this new 'ML job saved object'.

We haven't had to deal with migrating custom URLs on upgrade to date - the format of the URLs seems to have fairly stable over 6.x and 7.x. But if we were to switch to this new service, and storing the generator ID and state on the job, we'd need to add custom upgrade functionality to check all the job custom URLs and convert any old style full URL ones to the new generator ID / state format.

Reply:
Thanks for the details @peteharverson.  All that being said, is this specific change, in this PR, something you would like to see merged? There is much room for improvement with more refactoring/more registration of url generators, like discover, but for this initial PR, is this something you see as an improvement over the old code?  Or would you prefer to wait until you can implement a larger refactoring yourself?